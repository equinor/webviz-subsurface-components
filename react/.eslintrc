{
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 6,
    "sourceType": "module"
  },
  "extends":  [
    "plugin:react/recommended",
    "eslint:recommended",
    "plugin:prettier/recommended"
  ],
  "overrides": [
    {
      "files": ["*.ts", "*.tsx"],
      "plugins": ["@typescript-eslint"],
      "extends": [
        "plugin:react/recommended",
        "plugin:@typescript-eslint/recommended",
        "plugin:prettier/recommended"
      ]
    }
  ],
  "settings": {
    "react": {
      "version": "detect"
    }
  },
  "env": {
    "amd": true,
    "browser": true,
    "es6": true,
    "jasmine": true,
    "node": true
  }
}
