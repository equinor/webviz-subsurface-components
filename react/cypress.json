{
    "baseUrl": "http://localhost:6006/?path=/story/deckglmap--default",
    "video": true
}
