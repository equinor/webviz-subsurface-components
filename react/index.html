<html>
    <head>
        <title>Demo application of webviz-subsurface-components</title>
    </head>
    <body>
        <div id='root'></div>
        <link rel="stylesheet" type="text/css" href="output.css">
        <script src="output.js"></script>
    </body>
</html>
