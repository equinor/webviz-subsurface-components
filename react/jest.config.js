/**
 * The jest tests had an error: SyntaxError: Cannot use import statement outside a module
 * After some investigation, it leads to this page https://github.com/facebook/jest/issues/9395.
 */
module.exports = {
    moduleNameMapper: {
        "\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$":
            "<rootDir>/__mocks__/fileMock.js",
        "\\.(css|less)$": "<rootDir>/__mocks__/styleMock.js",
    },
    transform: {
        "\\.(js|ts|jsx|tsx)$": [
            "babel-jest",
            { configFile: "./config/babel.config.json" },
        ],
    },
    transformIgnorePatterns: ["<rootDir>/node_modules/(?!(@webviz)/)"],
};
