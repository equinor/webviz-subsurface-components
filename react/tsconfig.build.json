{
    "extends": "./tsconfig.json",
    "exclude": ["**/storybook/", "**/test/", "**/*.test.tsx"]
}
