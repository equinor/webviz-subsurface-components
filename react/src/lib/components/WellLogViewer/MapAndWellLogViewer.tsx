import React from "react";
import { WeakValidationMap } from "react";
import DeckGLMap from "../DeckGLMap";
import { DeckGLMapProps } from "../DeckGLMap";
import { Color } from "@deck.gl/core/typed";

import { WellsLayer } from "../DeckGLMap/layers";
import {
    Template,
    TemplateTrack,
    TemplatePlot,
    TemplatePlotTypes,
} from "./components/WellLogTemplateTypes";

import { WellLog } from "./components/WellLogTypes";
const welllogs =
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    require("../../../demo/example-data/volve_logs.json") as WellLog[];

const template =
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    require("../../../demo/example-data/welllog_template_2.json") as Template;

import { ColorTable } from "./components/ColorTableTypes";

import { WellLogController } from "./components/WellLogView";
import { LogViewer } from "@equinor/videx-wellog";
import { Info } from "./components/InfoTypes";
import { MapMouseEvent } from "../DeckGLMap/components/Map";

import InfoPanel from "./components/InfoPanel";
import WellLogViewWithScroller from "./components/WellLogViewWithScroller";
import { axisTitles, axisMnemos } from "./utils/axes";
import { fillInfos } from "./utils/fill-info";
import { getDiscreteMeta, indexOfElementByName } from "./utils/tracks";
import { deepCopy } from "./utils/tracks";

function getTemplatePlotColorTable(
    template: Template,
    templatePlot: TemplatePlot
) {
    let colorTable = templatePlot.colorTable;
    if (!colorTable && templatePlot.style) {
        const templateStyles = template.styles;
        if (templateStyles) {
            const iStyle = indexOfElementByName(
                templateStyles,
                templatePlot.style
            );
            if (iStyle >= 0) {
                const style = templateStyles[iStyle];
                colorTable = style.colorTable;
            }
        }
    }
    return colorTable;
}

type Props = DeckGLMapProps;

interface State {
    wellIndex: number | undefined;
    infos: Info[];
    controller?: WellLogController;
    editedData?: Record<string, unknown>;

    layers?: Record<string, unknown>[];

    wellName?: string;
    selection?: [number | undefined, number | undefined];
    selPersistent?: boolean;
    wellColor?: Color; // well color
}

function findWellsLayer(event: MapMouseEvent) {
    const info = event.infos.find((info) => info.layer?.id === "wells-layer");
    return info?.layer as WellsLayer;
}

function findWellLogIndex(welllogs: WellLog[], wellName: string): number {
    return welllogs.findIndex((welllog) => welllog.header.well === wellName);
}

function findLog(template: Template, logName: string): number {
    return template.tracks.findIndex(
        (track) => track.plots[0]?.name === logName
    );
}

function detectType(welllog: WellLog, logName: string): TemplatePlotTypes {
    if (welllog) {
        const meta = getDiscreteMeta(welllog, logName); // non-standard extention of WellLog JSON file
        if (meta) return "stacked";
    }
    return "line";
}

function addTemplateTrack(
    template: Template,
    welllog: WellLog,
    logName: string
): Template {
    // add missed TemplateTrack for the given logName
    const type: TemplatePlotTypes = detectType(welllog, logName);
    const templateNew = deepCopy(template);
    const templateTrack: TemplateTrack = {
        title: logName,
        required: true, // force to show on all wells
        plots: [{ name: logName, type: type, color: "red" }],
    };
    templateNew.tracks.push(templateTrack);
    return templateNew;
}

const wellpick = {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    wellpick: require("../../../demo/example-data/wellpicks.json")[0],
    name: "HORIZON",
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    colorTables: require("../../../demo/example-data/wellpick_colors.json"),
    color: "Stratigraphy",
};

export class MapAndWellLogViewer extends React.Component<Props, State> {
    public static propTypes?: WeakValidationMap<Props> | undefined;
    constructor(props: Props, state: State) {
        super(props, state);
        this.state = {
            wellIndex: undefined,
            infos: [],
            editedData: props.editedData,
            layers: props.layers,
        };
        this.onInfo = this.onInfo.bind(this);
        this.onCreateController = this.onCreateController.bind(this);
        this.onContentSelection = this.onContentSelection.bind(this);
        this.onTrackScroll = this.onTrackScroll.bind(this);

        this.onMouseEvent = this.onMouseEvent.bind(this);
    }
    componentDidUpdate(prevProps: Props /*, prevState: State*/): void {
        if (this.props.editedData !== prevProps.editedData) {
            this.setState({ editedData: this.props.editedData });
            0;
        }
    }
    onInfo(
        x: number,
        logController: LogViewer,
        iFrom: number,
        iTo: number
    ): void {
        const infos = fillInfos(
            x,
            logController,
            iFrom,
            iTo,
            [] //this.collapsedTrackIds,
            //this.props.readoutOptions
        );

        this.setState({ infos: infos });
    }

    onCreateController(controller: WellLogController): void {
        this.setState({ controller: controller });
    }
    onContentSelection(): void {
        const controller = this.state.controller;
        if (!controller) return;
        const selection = controller.getContentSelection();

        // synchronize selection only from the current well
        /*if (?? === this.state.wellName)*/ {
            this.setState({
                selection: selection,
                selPersistent: selection[1] !== undefined,
            });
        }
    }
    onTrackScroll(): void {
        const controller = this.state.controller;
        if (!controller) return;
        const iTrack = controller.getTrackScrollPos();
        if (iTrack >= 0) {
            const template = controller.getTemplate();
            const track = template.tracks[iTrack];
            if (track) {
                const templatePlot = track.plots[0];
                if (templatePlot) {
                    const wells_layer = this.props.layers?.find(
                        (item) => item["@@type"] === "WellsLayer"
                    );
                    if (
                        wells_layer &&
                        wells_layer["logName"] !== templatePlot.name
                    ) {
                        wells_layer["logName"] = templatePlot.name;
                        const colorTable = getTemplatePlotColorTable(
                            template,
                            templatePlot
                        );
                        if (colorTable) wells_layer["logColor"] = colorTable;
                        //(wells_layer.context as DeckGLLayerContext).userData.colorTables=colorTables;

                        const layers = deepCopy(this.props.layers);
                        this.setState({ layers: layers });

                        // Force to rerender ColorLegend after
                        setTimeout(() => {
                            const layers = deepCopy(this.props.layers);
                            this.setState({ layers: layers });
                        }, 200);
                    }
                }
            }
        }
    }

    onMouseEvent(event: MapMouseEvent): void {
        if (event.wellname !== undefined) {
            if (event.type == "click") {
                const iWell = findWellLogIndex(welllogs, event.wellname);
                this.setState((state: Readonly<State>) => {
                    //if (state.wellIndex === iWell) return null;

                    let selection:
                        | [number | undefined, number | undefined]
                        | undefined = undefined;
                    let selPersistent: boolean | undefined = undefined;
                    if (
                        state.wellIndex !== iWell ||
                        !state.selection ||
                        state.selPersistent
                    ) {
                        selection = [event.md, undefined];
                        selPersistent = false;
                    } else {
                        if (state.selection[1] !== undefined) {
                            // have something pinned
                            selection = [event.md, state.selection[1]];
                            selPersistent = true;
                        } else {
                            // no pinned yet
                            selection = [event.md, state.selection[0]]; // copy current to pinned
                            selPersistent = false;
                        }
                    }

                    return {
                        wellIndex: iWell,
                        wellName: event.wellname,
                        wellColor: event.wellcolor,
                        selection: selection,
                        selPersistent: selPersistent,
                    };
                });

                const controller = this.state.controller;
                if (controller) {
                    const wellsLayer = findWellsLayer(event);
                    if (wellsLayer) {
                        const template = controller.getTemplate();
                        const logName = wellsLayer.props?.logName;
                        let iTrack = findLog(template, logName);
                        if (iTrack < 0) {
                            //const welllog = info.object is Feature or WellLog;
                            const welllog = welllogs[iWell];
                            const templateNew = addTemplateTrack(
                                template,
                                welllog,
                                logName
                            );
                            controller.setTemplate(templateNew);

                            iTrack = findLog(template, logName);
                        }
                        controller.scrollTrackTo(iTrack);
                    }
                }
            }
            if (event.wellname === this.state.wellName) {
                // synchronize selection only from the current well
                if (event.md !== undefined) {
                    this.setState((state: Readonly<State>) => {
                        if (state.selPersistent) return null;

                        this.state.controller?.selectContent([
                            event.md,
                            this.state.selection?.[1],
                        ]);

                        return {
                            selection: [event.md, state.selection?.[1]],
                        };
                    });

                    //if (wellsLayer)
                    //    wellsLayer.setSelection(event.wellname, [event.md, undefined]);
                }
            }
        }
    }

    render(): JSX.Element {
        const wellName = this.state.wellName;
        const wellColor = this.state.wellColor;
        const wellIndex = this.state.wellIndex;
        return (
            <div style={{ height: "100%", width: "100%", display: "flex" }}>
                <div
                    style={{
                        height: "100%",
                        width: "70%",
                        position: "relative",
                    }}
                >
                    <div>
                        <DeckGLMap
                            {...this.props}
                            layers={this.state.layers}
                            editedData={this.state.editedData}
                            onMouseEvent={this.onMouseEvent}
                            selection={{
                                well: wellName,
                                selection: this.state.selection,
                            }}
                        />
                    </div>
                </div>
                <div
                    style={{
                        height: "85%",
                        width: "30%",
                        display: "flex",
                        flexDirection: "column",
                    }}
                >
                    <div
                        style={{
                            textAlign: "center",
                            flex: "0 0",
                        }}
                    >
                        {wellColor && (
                            <span
                                style={{
                                    color: wellColor
                                        ? "rgb(" +
                                          wellColor[0] +
                                          "," +
                                          wellColor[1] +
                                          "," +
                                          wellColor[2] +
                                          ")"
                                        : undefined,
                                    fontSize: "small",
                                }}
                            >
                                {"\u2B24 " /*big circle*/}
                            </span>
                        )}

                        {wellName
                            ? wellName
                            : "Select well by clicking on the map"}
                    </div>
                    <div
                        style={{
                            flex: "1 1",
                            height: "90%",
                            minWidth: "25px",
                            width: "100%",
                        }}
                    >
                        <div className="welllogview-error">
                            {wellIndex === -1
                                ? "No well logs for the well '" + wellName + "'"
                                : ""}
                        </div>
                        <WellLogViewWithScroller
                            welllog={
                                wellIndex !== undefined
                                    ? welllogs[wellIndex]
                                    : undefined
                            }
                            template={template}
                            colorTables={this.props.colorTables as ColorTable[]}
                            wellpick={wellpick}
                            maxVisibleTrackNum={1}
                            primaryAxis={"md"}
                            axisTitles={axisTitles}
                            axisMnemos={axisMnemos}
                            onInfo={this.onInfo}
                            onCreateController={this.onCreateController}
                            onContentSelection={this.onContentSelection}
                            onTrackScroll={this.onTrackScroll}
                        />
                    </div>
                    <div
                        style={{
                            flex: "0 0",
                            display: "flex",
                            flexDirection: "column",
                            height: "100%",
                            width: "100%",
                        }}
                    >
                        <InfoPanel header="Readout" infos={this.state.infos} />
                    </div>
                </div>
            </div>
        );
    }
}

MapAndWellLogViewer.propTypes = { ...DeckGLMap.propTypes };
