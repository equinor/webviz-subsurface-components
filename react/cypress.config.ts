import { defineConfig } from "cypress";
import getCompareSnapshotsPlugin from "cypress-image-diff-js/dist/plugin";

export default defineConfig({
  component: {
    devServer: {
      framework: "create-react-app",
      bundler: "webpack",
    },
    video:false,
    setupNodeEvents(on, config) {
      getCompareSnapshotsPlugin(on, config),
      on('before:browser:launch', (browser = {}, launchOptions) => {
        if (browser.family === 'chromium' && browser.name !== 'electron') {
          launchOptions.args.push('--start-fullscreen')
      
          return launchOptions
        }
        if (browser.name === 'electron') {
          launchOptions.preferences.fullscreen = true
      
          return launchOptions
        }
      })
    }
  },
});
