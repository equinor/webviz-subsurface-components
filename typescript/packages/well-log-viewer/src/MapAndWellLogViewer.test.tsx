/* eslint-disable @typescript-eslint/no-var-requires */
import "@testing-library/jest-dom/extend-expect";
import { render } from "@testing-library/react";
import "jest-styled-components";
import React from "react";
import { MapAndWellLogViewer } from "./MapAndWellLogViewer";

window.ResizeObserver =
    window.ResizeObserver ||
    jest.fn().mockImplementation(() => ({
        disconnect: jest.fn(),
        observe: jest.fn(),
        unobserve: jest.fn(),
    }));

describe("Map and Well Log Viewer", () => {
    it("snapshot test", () => {
        const { container } = render(
            <MapAndWellLogViewer
                id={""}
                bounds={[432205, 6475078, 437720, 6481113]}
            />
        );
        expect(container.firstChild).toMatchSnapshot();
    });
});
