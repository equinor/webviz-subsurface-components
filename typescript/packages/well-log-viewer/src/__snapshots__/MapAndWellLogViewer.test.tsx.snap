// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map and Well Log Viewer snapshot test 1`] = `
.c0 {
  display: inline-block;
  -webkit-transform: rotate(-90deg);
  -ms-transform: rotate(-90deg);
  transform: rotate(-90deg);
}

<div
  style="height: 100%; width: 100%; display: flex;"
>
  <div
    style="height: 100%; width: 70%; position: relative;"
  >
    <div>
      <div>
        <div
          id="deckgl-wrapper"
          style="position: absolute; z-index: 0; left: 0px; top: 0px; width: 100%; height: 100%;"
        >
          <canvas
            id="deckgl-overlay"
            style="left: 0px; top: 0px; width: 100%; position: absolute; height: 100%;"
          />
        </div>
        <div>
          <svg
            aria-valuemax="100"
            aria-valuemin="0"
            aria-valuenow="0"
            class="c0"
            height="48"
            preserveAspectRatio="xMidYMid meet"
            role="progressbar"
            viewBox="24 24 48 48"
            width="48"
          >
            <circle
              class=""
              cx="48"
              cy="48"
              fill="none"
              r="22"
              stroke="var(--eds_infographic_primary__moss_green_13, rgba(222, 237, 238, 1))"
              stroke-width="4"
              style="stroke: 138.230; stroke-dashoffset: 138.230px;"
            />
            <circle
              class=""
              cx="48"
              cy="48"
              fill="none"
              r="22"
              stroke="var(--eds_infographic_primary__moss_green_100, rgba(0, 112, 121, 1))"
              stroke-dasharray="138.23007675795088"
              stroke-linecap="round"
              stroke-width="4"
              style="stroke: 138.230; stroke-dashoffset: 138.230px;"
            />
          </svg>
          <br />
          Loading assets...
        </div>
      </div>
    </div>
  </div>
  <div
    style="height: 85%; width: 30%; display: flex; flex-direction: column;"
  >
    <div
      style="flex: 1 1; height: 90%; min-width: 25px; width: 100%;"
    >
      <div
        style="overflow: scroll; width: 100%; height: 100%;"
      >
        <div
          style="width: 0px;"
        >
          <div
            style="position: absolute;"
          >
            <div
              class="welllogview"
              style="width: 100%; height: 100%; display: flex; flex-direction: column;"
            >
              <div
                class="title"
              >
                <div
                  style="font-size: 16px;"
                >
                  Select a well by clicking on the map
                </div>
              </div>
              <div
                style="width: 100%; height: 100%; display: flex; flex-direction: column;"
              >
                <div
                  class="log-controller"
                >
                  <div
                    class="overlay"
                    style="width: 0px; left: 0px; overflow: hidden;"
                  >
                    <div
                      class="depth"
                      style="position: absolute; pointer-events: none; visibility: hidden;"
                    />
                    <div
                      class="rubber-band"
                      style="position: relative; pointer-events: none; height: 9px; width: 100%; visibility: hidden;"
                    >
                      <div
                        style="height: 1px; width: 100%; top: 4.5px; position: relative;"
                      />
                    </div>
                    <div
                      class="pinned"
                      style="position: absolute; pointer-events: none; height: 9px; width: 100%; left: 0px; visibility: hidden;"
                    >
                      <div
                        style="height: 1px; width: 100%; top: 4.5px; position: absolute;"
                      />
                    </div>
                    <div
                      class="wellpick"
                      style="position: absolute; pointer-events: none; height: 3px; width: 100%; left: 0px; background-color: rgba(255, 193, 0, 0.8); visibility: false;"
                    >
                      <div
                        style="position: absolute; height: 1px; width: 100%; background-color: rgba(255, 193, 0, 0.8);"
                      >
                        <table
                          style="position:relative; top:-1.5em;"
                          width="100%"
                        >
                          <tbody>
                            <tr>
                              <td
                                class="wp-title-0"
                              >
                                <span
                                  style="background-color:rgba(255,193,0,0.16)"
                                >
                                  1644
                                </span>
                              </td>
                              <td
                                class="wp-title-1"
                              >
                                <span
                                  style="background-color:rgba(255,193,0,0.16)"
                                />
                              </td>
                              <td
                                align="center"
                              >
                                <span
                                  style="background-color:rgba(255,193,0,0.16)"
                                >
                                  Hor_1
                                </span>
                              </td>
                            </tr>
                          </tbody>
                        </table>
                      </div>
                    </div>
                    <div
                      class="wellpick"
                      style="position: absolute; pointer-events: none; height: 3px; width: 100%; left: 0px; background-color: rgba(255, 120, 61, 0.8); visibility: false;"
                    >
                      <div
                        style="position: absolute; height: 1px; width: 100%; background-color: rgba(255, 120, 61, 0.8);"
                      >
                        <table
                          style="position:relative; top:-1.5em;"
                          width="100%"
                        >
                          <tbody>
                            <tr>
                              <td
                                class="wp-title-0"
                              >
                                <span
                                  style="background-color:rgba(255,120,61,0.16)"
                                >
                                  2788
                                </span>
                              </td>
                              <td
                                class="wp-title-1"
                              >
                                <span
                                  style="background-color:rgba(255,120,61,0.16)"
                                />
                              </td>
                              <td
                                align="center"
                              >
                                <span
                                  style="background-color:rgba(255,120,61,0.16)"
                                >
                                  Hor_2
                                </span>
                              </td>
                            </tr>
                          </tbody>
                        </table>
                      </div>
                    </div>
                    <div
                      class="wellpick"
                      style="position: absolute; pointer-events: none; height: 3px; width: 100%; left: 0px; background-color: rgba(255, 155, 76, 0.8); visibility: false;"
                    >
                      <div
                        style="position: absolute; height: 1px; width: 100%; background-color: rgba(255, 155, 76, 0.8);"
                      >
                        <table
                          style="position:relative; top:-1.5em;"
                          width="100%"
                        >
                          <tbody>
                            <tr>
                              <td
                                class="wp-title-0"
                              >
                                <span
                                  style="background-color:rgba(255,155,76,0.16)"
                                >
                                  3288
                                </span>
                              </td>
                              <td
                                class="wp-title-1"
                              >
                                <span
                                  style="background-color:rgba(255,155,76,0.16)"
                                />
                              </td>
                              <td
                                align="center"
                              >
                                <span
                                  style="background-color:rgba(255,155,76,0.16)"
                                >
                                  Hor_3
                                </span>
                              </td>
                            </tr>
                          </tbody>
                        </table>
                      </div>
                    </div>
                    <div
                      class="wellpick"
                      style="position: absolute; pointer-events: none; height: 3px; width: 100%; left: 0px; background-color: rgba(255, 223, 161, 0.8); visibility: false;"
                    >
                      <div
                        style="position: absolute; height: 1px; width: 100%; background-color: rgba(255, 223, 161, 0.8);"
                      >
                        <table
                          style="position:relative; top:-1.5em;"
                          width="100%"
                        >
                          <tbody>
                            <tr>
                              <td
                                class="wp-title-0"
                              >
                                <span
                                  style="background-color:rgba(255,223,161,0.16)"
                                >
                                  3611
                                </span>
                              </td>
                              <td
                                class="wp-title-1"
                              >
                                <span
                                  style="background-color:rgba(255,223,161,0.16)"
                                />
                              </td>
                              <td
                                align="center"
                              >
                                <span
                                  style="background-color:rgba(255,223,161,0.16)"
                                >
                                  Hor_4
                                </span>
                              </td>
                            </tr>
                          </tbody>
                        </table>
                      </div>
                    </div>
                    <div
                      class="wellpick"
                      style="position: absolute; pointer-events: none; height: 3px; width: 100%; left: 0px; background-color: rgba(255, 223, 161, 0.8); visibility: false;"
                    >
                      <div
                        style="position: absolute; height: 1px; width: 100%; background-color: rgba(255, 223, 161, 0.8);"
                      >
                        <table
                          style="position:relative; top:-1.5em;"
                          width="100%"
                        >
                          <tbody>
                            <tr>
                              <td
                                class="wp-title-0"
                              >
                                <span
                                  style="background-color:rgba(255,223,161,0.16)"
                                >
                                  3911
                                </span>
                              </td>
                              <td
                                class="wp-title-1"
                              >
                                <span
                                  style="background-color:rgba(255,223,161,0.16)"
                                />
                              </td>
                              <td
                                align="center"
                              >
                                <span
                                  style="background-color:rgba(255,223,161,0.16)"
                                >
                                  Hor_5
                                </span>
                              </td>
                            </tr>
                          </tbody>
                        </table>
                      </div>
                    </div>
                    <div
                      class="wellpick"
                      style="position: absolute; pointer-events: none; height: 3px; width: 100%; left: 0px; background-color: rgba(255, 223, 161, 0.8); visibility: false;"
                    >
                      <div
                        style="position: absolute; height: 1px; width: 100%; background-color: rgba(255, 223, 161, 0.8);"
                      >
                        <table
                          style="position:relative; top:-1.5em;"
                          width="100%"
                        >
                          <tbody>
                            <tr>
                              <td
                                class="wp-title-0"
                              >
                                <span
                                  style="background-color:rgba(255,223,161,0.16)"
                                >
                                  4411
                                </span>
                              </td>
                              <td
                                class="wp-title-1"
                              >
                                <span
                                  style="background-color:rgba(255,223,161,0.16)"
                                />
                              </td>
                              <td
                                align="center"
                              >
                                <span
                                  style="background-color:rgba(255,223,161,0.16)"
                                >
                                  Hor_6
                                </span>
                              </td>
                            </tr>
                          </tbody>
                        </table>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      style="flex: 0 0 0px; display: flex; flex-direction: column; height: 100%; width: 100%;"
    >
      <div
        style="overflow-y: auto; overflow-x: hidden;"
      >
        <fieldset>
          <legend>
            Readout
          </legend>
          <table
            style="border-spacing: 0px; width: 100%;"
          >
            <tbody>
              <tr>
                <td
                  colspan="4"
                >
                   
                  <hr />
                </td>
              </tr>
            </tbody>
          </table>
        </fieldset>
      </div>
    </div>
  </div>
</div>
`;
